### PR TITLE
Micro optimizations, minor refactories and QOL 

### PR DESCRIPTION
**Pull request by xGab0**
- migrate from Java Collection to FastUtil Collection
- record convertion for the acoustic system

### CODE REVIEW

**Comment by Sollace** [comment]
Sorry, I did not get to see this before doing the release.

File: src/main/java/eu/ha3/presencefootsteps/sound/Options.java [hunk 1]
```diff
@@ -30,4 +31,5 @@ public float get(String option) {
     boolean containsKey(String option);
 
     float get(String option);
-}
+
+}
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/sound/Options.java:35]: Can you please go through and remove whitespace changes like these.

File: src/main/java/eu/ha3/presencefootsteps/sound/acoustics/AcousticsPlayer.java [hunk 2]
```diff
@@ -1,23 +1,24 @@
...
-public class AcousticsPlayer implements AcousticLibrary {
-
-    private final SoundPlayer player;
+import java.util.Map;
+import java.util.stream.Stream;
 
-    private final Map<String, Acoustic> acoustics = new HashMap<>();
+public record AcousticsPlayer(
+        @NotNull SoundPlayer player,
+        @NotNull Map<String, Acoustic> acoustics
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/sound/acoustics/AcousticsPlayer.java:17]: You can remove the @ NotNull annotations. I instead annotate things with @ Nullable if they may be null

File: src/main/java/eu/ha3/presencefootsteps/sound/acoustics/ChanceAcoustic.java [hunk 3]
```diff
@@ -5,30 +5,33 @@
...
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Range;
+
+record ChanceAcoustic(
+        @NotNull Acoustic acoustic,
+        @Range(from = 1, to = Integer.MAX_VALUE) float probability
+) implements Acoustic {
+
+    @Contract(value = "_, _ -> new", pure = true)
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/sound/acoustics/ChanceAcoustic.java:17]: Remove @ Range and @ Contract annotations. Not sure what purpose they're going to serve as nothing I have makes use of them.

File: src/main/java/eu/ha3/presencefootsteps/sound/acoustics/VaryingAcoustic.java [hunk 4]
```diff
@@ -45,16 +61,15 @@ public void playSound(SoundPlayer player, LivingEntity location, State event, Op
...
-        }
-        if (inputOptions.containsKey("gliding_pitch")) {
-            pitch = this.pitch.on(inputOptions.get("gliding_pitch"));
-        }
+        final float pitch = inputOptions.containsKey("gliding_pitch")
+                ? this.pitch.on(inputOptions.get("gliding_pitch"))
+                : this.pitch.random(player.getRNG());
 
-        player.playSound(location, soundName, volume, pitch, getOptions());
+        player.playSound(location, soundName, volume, pitch, Options.EMPTY);
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/sound/acoustics/VaryingAcoustic.java:72]: Why is this changed to Options.EMPTY here rather than calling getOptions()?

File: src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java [hunk 5]
```diff
@@ -1,19 +1,20 @@
...
+import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap;
 import net.minecraft.entity.EntityType;
 import net.minecraft.util.Identifier;
 
-public class GolemLookup implements Lookup<EntityType<?>> {
+import java.util.Map;
+import java.util.Set;
 
-    private final Map<String, Map<Identifier, String>> substrates = new LinkedHashMap<>();
+public record GolemLookup(Map<String, Map<Identifier, String>> substrates) implements Lookup<EntityType<?>> {
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java:10]: Not everything has to be a record

File: src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java [hunk 6]
```diff
@@ -1,26 +1,27 @@
...
 
     @Override
     public Locomotion lookup(Entity key) {
-        if (key instanceof PlayerEntity) {
-            return Locomotion.forPlayer((PlayerEntity)key, Locomotion.NONE);
-        }
-        return Locomotion.forLiving(key, values.getOrDefault(EntityType.getId(key.getType()), Locomotion.BIPED));
+        return key instanceof PlayerEntity player
+                ? Locomotion.forPlayer(player, Locomotion.NONE)
+                : Locomotion.forLiving(key, values.getOrDefault(EntityType.getId(key.getType()), Locomotion.BIPED));
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java:24]: Revert this please for readability.

**Review by Sollace — CHANGES_REQUESTED** [review]

File: src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java (hunk 5 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java:10]: Records are usually faster than normal classes and they should be used when possible

File: src/main/java/eu/ha3/presencefootsteps/sound/acoustics/ChanceAcoustic.java (hunk 3 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/sound/acoustics/ChanceAcoustic.java:17]: I removed the Range annotation but for the Contract annotation I'm against removing it, I've pasted this from the docs:
_"Specifies some aspects of the method behavior depending on the arguments. Can be used by tools for advanced data flow analysis. Note that this annotation just describes how the code works and doesn't add any functionality by means of code generation."_

File: src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java (hunk 6 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java:24]: Is this really that bad? It looks very comprehensible to me: the ternary operator is at the method return and shouldn't confuse that much.

File: src/main/java/eu/ha3/presencefootsteps/sound/acoustics/VaryingAcoustic.java (hunk 4 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/sound/acoustics/VaryingAcoustic.java]: I removed the #getOptions method while converting VaryingAcoustic & DelayedAcoustic to Records since VaryingAcoustic forced Options#EMPTY anyway. I've now rewritten the #playSound method to share code between the two classes like before.

File: src/main/java/eu/ha3/presencefootsteps/sound/Options.java (hunk 1 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/sound/Options.java:33]: I should've removed all the whitespaces in the new commit

File: src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java (hunk 5 above)
Comment (Sollace) [reply src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java:10]: Records are normal classes with an implicit all-args-constructor and overrides for equals(), hashCode(), and toString(), none of which are particularly required for this class. This class has one field and it's assigned immediately at construction, which I generally do not use records for.

File: src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java (hunk 6 above)
Comment (Sollace) [reply src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java:24]: Maybe for you, but for me I'd rather have a separate if statement for this kind of thing. I also want to ensure the PR is lean before I merge as it makes it easier for me to know what's affected.

File: src/main/java/eu/ha3/presencefootsteps/sound/Options.java (hunk 1 above)
Comment (Sollace) [reply src/main/java/eu/ha3/presencefootsteps/sound/Options.java:33]: Thanks. This does look better. There's just a few whitespace changes that are still outstanding (this one and some indentations newline braces)

File: src/main/java/eu/ha3/presencefootsteps/sound/SoundEngine.java [hunk 7]
```diff
@@ -174,7 +173,8 @@ public Identifier getFabricId() {
     @Override
     public CompletableFuture<Void> reload(Synchronizer sync, ResourceManager sender,
             Profiler serverProfiler, Profiler clientProfiler,
-            Executor serverExecutor, Executor clientExecutor) {
+            Executor serverExecutor, Executor clientExecutor)
+    {
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/sound/SoundEngine.java:177]: Brace should be on the same line, please, and restore the terminating newline at the end of this file.

File: src/main/java/eu/ha3/presencefootsteps/sound/acoustics/AcousticsPlayer.java [hunk 8]
```diff
@@ -28,16 +29,15 @@ public void addAcoustic(String name, Acoustic acoustic) {
...
+                    .distinct()
+                    .forEach(fragment -> playAcoustic(location, fragment, event, inputOptions));
         } else if (!acoustics.containsKey(acousticName)) {
             PresenceFootsteps.logger.warn("Tried to play a missing acoustic: " + acousticName);
         } else {
             acoustics.get(acousticName).playSound(player, location, event, inputOptions);
         }
     }
-}
+}
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/sound/acoustics/AcousticsPlayer.java:43]: Same as before, please restore the newline.

File: src/main/java/eu/ha3/presencefootsteps/sound/acoustics/ChanceAcoustic.java [hunk 9]
```diff
@@ -5,30 +5,31 @@
...
         }
     }
-
-    public static Acoustic fromJson(JsonObject json, AcousticsJsonParser context) {
-        Acoustic acoustic = context.solveAcoustic(json.get("acoustic"));
-        float probability = json.get("probability").getAsFloat();
-        return new ChanceAcoustic(acoustic, probability);
-    }
-}
+}
```
Comment (Sollace) [inline src/main/java/eu/ha3/presencefootsteps/sound/acoustics/ChanceAcoustic.java:35]: Newline please

Comment (Sollace) [reply src/main/java/eu/ha3/presencefootsteps/sound/acoustics/ChanceAcoustic.java:35]: Not going to point out each one, just please check through the commit for the others. If there's files without terminating newlines already, it's a style issue on my end that I'll fix as I see them.

File: src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java (hunk 6 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/world/LocomotionLookup.java:24]: Fine, I'm going to revert this.

File: src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java (hunk 5 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/world/GolemLookup.java:10]: I understand, I'm going to revert this👍

File: src/main/java/eu/ha3/presencefootsteps/sound/Options.java (hunk 1 above)
Comment (xGab0) [reply src/main/java/eu/ha3/presencefootsteps/sound/Options.java:33]: I'm going to find the last whitespaces hoping I won't miss anything this time.

**Comment by Sollace** [comment]
We are expecting 1.20 to drop soon, so if you could have this ready before I start updating to it proper, I would love that.

**Comment by xGab0** [comment]
> We are expecting 1.20 to drop soon, so if you could have this ready before I start updating to it proper, I would love that.

Sure, I'm going to work on this as soon as I can

**Comment by Sollace** [comment]
Also no need to worry about conflicts. I'll sort those out when I merge. :)

**Review by Sollace — APPROVED** [review]

**Comment by xGab0** [comment]
Thank you for your patience, everything should be fine now

**Comment by Sollace** [comment]
Alright, thanks very much!
